### PR TITLE
PR: Correctly handle Kite replacement range, fixing dict completions

### PR DESCRIPTION
**Pull request by metalogical**
## Description of Changes

* [ ] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Kite returns a replacement range with each completion indicating where to insert the completion within the buffer. This is particularly useful for more non-traditional completions, such as Kite's dict completions, where simply matching the preceding identifier is not enough. Kite's dict completions were previously broken in Spyder, but this change fixes them.

1. Support the `textEdit` functionality of the [LSP spec](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#textDocument_completion)
2. Properly respect the `filterText` when deciding whether completions are still valid.
3. Update the Kite provider to specify a `textEdit` instead of `insertText`, and also not use `filterText`, since Kite filters internally (and will always do so -- it is part of the API spec).

Should I add any unit tests for this? @andfoy @ccordoba12 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @metalogical

### CODE REVIEW

**Comment by metalogical** [comment]
@andfoy I have not tested with other LSP servers. I'm not sure which server uses the `textEdit` functionality instead of `insertText`, so it's a little difficult to test. Any suggestions?

**Comment by metalogical** [comment]
I fixed a bug that was caught by tests where when on-the-fly completions are turned off, the completion filtering was broken:

1. use the old insert logic for the old-style `insertText` completions
2. remove `textEdit` completions from the completion list when doing in-editor filtering, since it's difficult to correctly update the replacement range.
    - so if on-the-fly completions are turned off, you'll see Kite completions on tab or ctrl-space, but they'll disappear when you continue typing.

**Comment by metalogical** [comment]
@andfoy I looked into the unit tests, and it seems like we already have fairly good coverage on tests that check that the PyLS (`insertText`) completions continue to work. So this should at least be fairly safe on that front.

We don't have much coverage on Kite completions (with `textEdit`), but I'm not sure if there's any good way to test that right now with your infrastructure.

**Comment by andfoy** [comment]
> We don't have much coverage on Kite completions (with textEdit), but I'm not sure if there's any good way to test that right now with your infrastructure.

Actually, right now we don't have any Kite tests, we can add them later on, so don't worry about that

**Comment by goanpeca** [comment]
Hi @metalogical thanks for working on this! Looks awesome :-)

I have one question. I was expecting to get completions without the need of typing the string quotes

```
dic = {}
dic['one'] = 1
dic['two'] = 2

dic[*]
```

So if I have the cursor (*) and press `tab` I just get a tab space and not a completion. Is this expected behavior?

File: spyder/plugins/editor/widgets/completion.py [hunk 1]
```diff
@@ -170,7 +170,7 @@ def show_list(self, completion_list, position, automatic):
         # signal used for testing
         self.sig_show_completions.emit(completion_list)
 
-    def update_list(self, filter_text):
+    def update_list(self, current_word, new=True):
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/completion.py]: @metalogical could you add more information in this docstring about `current_word` and `new` parameters?

File: spyder/plugins/editor/widgets/base.py [hunk 2]
```diff
@@ -854,11 +854,23 @@ def select_completion_list(self):
         """Completion list is active, Enter was just pressed"""
         self.completion_widget.item_selected()
 
-    def insert_completion(self, text, completion_position):
-        if text:
+    def insert_completion(self, completion, completion_position):
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/base.py]: Please add a docstring here

**Comment by metalogical** [comment]
@goanpeca thanks for the review.

There is currently a deficiency in Kite where we cannot parse incomplete subscript expressions, which is why you don't see completions in the `dic[*]` case.

It's a bit tricky for us to fix this right now, so no word on when there will be a fix, but it's in our queue.

I addressed your other comments.

**Comment by ccordoba12** [comment]
@metalogical, this has a conflict with master now (probably due to PR #10502). Please solve it so I can merge it.

**Comment by metalogical** [comment]
resolved the conflict, just waiting on CI

**Comment by goanpeca** [comment]
The following error is appearing @metalogical 

```python-traceback
completion = self.currentItem().data(Qt.UserRole)
AttributeError: 'NoneType' object has no attribute 'data'
```

**Comment by metalogical** [comment]
@ccordoba12 @goanpeca this should be good to go

**Pull request merged**